### ISSUE TITLE
'recurrent_dropout' with 'relu' in LSTM yields NaNs

### ISSUE THREAD

**Issue by OverLordGoldDragon** [keras-team/keras#13221]
Any non-zero `recurrent_dropout` yields NaN losses and weights; latter are either 0 or NaN. Happens for stacked, shallow, `stateful`, `return_sequences` = any, with & w/o `Bidirectional()`, `activation='relu'`, `loss='binary_crossentropy'`. NaNs occur within a few batches - the more layers, the sooner. 

Any fixes? Help's appreciated.

<hr>
<b>TROUBLESHOOTING ATTEMPTED</b>:

 - `recurrent_dropout=0.2,0.1,0.01,1e-6`
 - `kernel_constraint=maxnorm(0.5,axis=0)`
 - `recurrent_constraint=maxnorm(0.5,axis=0)`
 - `clipnorm=50`  (empirically determined), Nadam optimizer 
 - `activation='tanh'` - no NaNs, weights stable, tested for up to 10 batches
 - `lr=2e-6,2e-5` - no NaNs, weights stable, tested for up to 10 batches
 - `lr=5e-5` - no NaNs, weights stable, for 3 batches - NaNs on batch 4

_NOTE_: `batch_shape=(32,672,16)`, 17 calls to `train_on_batch` per batch

<hr>
<b>ENVIRONMENT</b>:

 - Keras 2.2.4 (TensorFlow backend), Python 3.7, Spyder 3.3.7 via Anaconda
 - GTX 1070 6GB, i7-7700HQ, 12GB RAM, Win-10.0.17134 x64
 - CuDNN 10+, latest Nvidia drives

**Comment by Rainymood** [comment]
Do you have a minimal reproducible example?

**Comment by OverLordGoldDragon** [comment]
@Rainymood 
```
from keras.constraints import maxnorm
constraint = maxnorm(0.5,axis=0)

ipt = Input(batch_shape=(32,1350,16))
x = LSTM(512,activation='relu',return_sequences=False,recurrent_dropout=0.1,
             kernel_constraint=constraint,recurrent_constraint=constraint)
out = Dense(1,activation='sigmoid')(x)

model = Model(ipt,out)
```
The actual model is much larger, with conv, batch norm, dropout, and pooling layers preceding LSTMs - but since their outputs are well-behaved, the `Input`  above should serve a reasonable substitute.

**Comment by OverLordGoldDragon** [comment]
Before I get a minimal reproducible code, some important results:

Model divergence is spontaneous, occurring at different train updates _even with fixed seeds_ - Numpy, Random, and TensorFlow random seeds. Furthermore, when first diverging, LSTM layer weights are all normal - only going to NaN later. 

Below are, in order: (1) inputs to `LSTM`; (2) `LSTM` outputs; (3) `Dense(1,'sigmoid')` outputs -- the three are consecutive, with `Dropout(0.5)` between each. Preceding (1) are `Conv1D` layers. Right: LSTM weights All are _one_ train update before divergence.

![image](https://user-images.githubusercontent.com/16495490/63199693-a9c7c580-c04c-11e9-9396-b77079352720.png)

_AT divergence_:
![image](https://user-images.githubusercontent.com/16495490/63199806-fca17d00-c04c-11e9-9224-be21a1c2d1c8.png)
```
## LSTM outputs, flattened, stats
(mean,std)        = (inf,nan)
(min,max)         = (0.00e+00,inf)
(abs_min,abs_max) = (0.00e+00,inf)
```
_AFTER divergence_:
![image](https://user-images.githubusercontent.com/16495490/63200017-9c5f0b00-c04d-11e9-87e5-9ace4cbf9407.png)
```
## Recurrent Gates Weights:
array([[nan, nan, nan, ..., nan, nan, nan],
       [ 0.,  0., -0., ..., -0.,  0.,  0.],
       [ 0., -0., -0., ..., -0.,  0.,  0.],
       ...,
       [nan, nan, nan, ..., nan, nan, nan],
       [ 0.,  0., -0., ..., -0.,  0., -0.],
       [ 0.,  0., -0., ..., -0.,  0.,  0.]], dtype=float32)
## Dense Sigmoid Outputs:
array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]], dtype=float32)
```

**Comment by OverLordGoldDragon** [comment]
**Minimal reproducible example**:

```
from keras.layers import Input,Dense,LSTM,Dropout
from keras.models import Model
from keras.optimizers  import Nadam 
from keras.constraints import maxnorm
```
```
ipt = Input(batch_shape=(32,672,16))
x = LSTM(512,activation='relu',return_sequences=False,
             recurrent_dropout=0.1,
             kernel_constraint   =maxnorm(0.5,axis=0),
             recurrent_constraint=maxnorm(0.5,axis=0))(ipt)
out = Dense(1,activation='sigmoid')(x)

model = Model(ipt,out)
optimizer = Nadam(lr=2e-4,clipnorm=1)
model.compile(optimizer=optimizer,loss='binary_crossentropy')
```
```
for train_update,_ in enumerate(range(100)):
    x = np.random.randn(32,672,16)
    y = np.array([1]*5 + [0]*27)
    np.random.shuffle(y)
    loss = model.train_on_batch(x,y)
    print(train_update+1,loss,np.sum(y))
```
**Observations**: the following _speed up divergence_:

- _Higher_ `lr`
- _Higher_ `units` (LSTM)
- _Higher_ # of layers (LSTM)
- _Less_ `'1'` labels **<<**

The last one is important: model doesn't diverge using `y` below even with `lr=1e-3`, for up to 400 train updates; `np.sum(y)` above shows # of `'1'` labels

`y = np.random.randint(0,2,32); np.random.shuffle(y)`

**Comment by OverLordGoldDragon** [comment]
Tested more activations, with one strange result:

**DIVERGENCE**:
- `relu`
- `elu`
- `linear`

**NO DIVERGENCE**:
- `tanh`
- `sigmoid`
- `selu`

```https://github.com/keras-team/keras/blob/master/keras/activations.py
SELU is equal to: `scale * elu(x, alpha)`, where alpha and scale
    are predefined constants. The values of `alpha` and `scale` are
    chosen so that the mean and variance of the inputs are preserved
    between two consecutive layers
```

**Comment by OverLordGoldDragon** [comment]
**UPDATE**: Does _NOT_ occur with GRUs. No clue why.

**Comment by OverLordGoldDragon** [comment]
_Update_: Happens with `activation='tanh'` also - simply spontaneously, don't even know how to reproduce.

**Comment by OverLordGoldDragon** [comment]
Any progress on this? I encountered a new unique instance: _validation_ NaNs, while _train_ is fine - bizarre case, don't know how to reproduce minimally

**Comment by OverLordGoldDragon** [comment]
[Solved](https://stackoverflow.com/questions/57516678/lstm-recurrent-dropout-with-relu-yields-nans/59656753#59656753).
